src/gt4py/next/ffront/foast_passes/type_deduction.py

```diff
@@ -546,6 +546,29 @@ def visit_Call(self, node: foast.Call, **kwargs) -> foast.Call:
         arg_types = [arg.type for arg in new_args]
         kwarg_types = {name: arg.type for name, arg in new_kwargs.items()}
 
+        func_str_repr: str  # just for error handling
+        if isinstance(
+            new_func.type,
+            (ts.FunctionType, ts_ffront.FieldOperatorType, ts_ffront.ScanOperatorType),
+        ):
+            # Since we use the `id` attribute in the latter part of the toolchain ensure we
+            # have the proper format here.
+            if not isinstance(
+                new_func,
+                (foast.FunctionDefinition, foast.FieldOperator, foast.ScanOperator, foast.Name),
+            ):
+                raise FieldOperatorTypeDeductionError.from_foast_node(
+                    node, msg="Functions can only be called directly!"
+                )
+            func_str_repr = new_func.id
+        elif isinstance(new_func.type, ts.FieldType):
+            func_str_repr = str(new_func)
+        else:
+            raise FieldOperatorTypeDeductionError.from_foast_node(
+                node,
+                msg=f"Expression of type `{new_func.type}` is not callable, must be a `Function`, `FieldOperator`, `ScanOperator` or `Field`.",
+            )
+
         # ensure signature is valid
         try:
             type_info.accepts_args(
```

Comment:
This snippet, including the `type_info.accepts_args` is confusing. How does that work in case of `new_func.type == FieldType`? Let's talk about it.